roachtest,asim: validate asim

kvoli opened the pull request:
This PR has some misc fixes to the allocator for non-determinism and local accounting following a rebalance action.

Some additional metrics are added for tracking store level simulated cluster state, as well as various smaller fixes implemented for validity issues.

A non-asserting roachtest is added for easy comparison of the real and simulated cluster results for validation. 


resolves: https://github.com/cockroachdb/cockroach/issues/89394
resolves: https://github.com/cockroachdb/cockroach/issues/89397

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/88068)
<!-- Reviewable:end -->

kvoli committed:
kvserver: fix allocator determinism

Previously, the allocator when running in deterministic mode, would not
be deterministic under all scenarios due to iteration over an unordered
map. This patch fixes this by first sorting entries, before iterating.

resolves: #89394

Release note: None

asim,kvserver: make simulator deterministic

Previously, after rebalancing a range via `RelocateRange`, the store
rebalancer would not update the local storepool to reflect the change.

This patch adds this update, so that decisions may be made more
accurately on the basis of recent prior actions. This also updates
the store rebalancer to "refresh" the rebalance context with the latest
storepool values, rather than maintain custom logic locally.

Additional metrics also added.

resolves: #89397

Release note: None

roachtest: add asim validation

This patch adds code to run equivalent configurations for a
real and simulated cluster. It parses the real cluster results and
converts it into the simulator format which is then written to a csv.

There are no assertions in any test and by default the real cluster is
not enabled to run, only the simulation.

The purpose of checking in this roachtest is to easily compare simulator
and the real cluster output, for validation.

Release note: None
